pydantic_ai_slim/pydantic_ai/ui/vercel_ai/_adapter.py

```diff
@@ -197,3 +231,204 @@ def load_messages(cls, messages: Sequence[UIMessage]) -> list[ModelMessage]:  #
+
+                if user_ui_parts:
+                    result.append(UIMessage(id=_message_id_generator(), role='user', parts=user_ui_parts))
+
+            elif isinstance(  # pragma: no branch
+                msg, ModelResponse
+            ):
+                ui_parts: list[UIMessagePart] = []
+
+                # For builtin tools, returns can be in the same ModelResponse as calls
+                # Build a local mapping for this message
+                local_builtin_returns: dict[str, BuiltinToolReturnPart] = {}
+                for part in msg.parts:
+                    if isinstance(part, BuiltinToolReturnPart):
+                        local_builtin_returns[part.tool_call_id] = part
+
+                for part in msg.parts:
+                    if isinstance(part, BuiltinToolReturnPart):
+                        continue
+                    elif isinstance(part, TextPart):
+                        # Combine consecutive text parts by checking the last UI part
+                        if ui_parts and isinstance(ui_parts[-1], TextUIPart):
+                            last_text = ui_parts[-1]
+                            ui_parts[-1] = last_text.model_copy(update={'text': last_text.text + part.content})
+                        else:
+                            ui_parts.append(TextUIPart(text=part.content, state='done'))
+                    elif isinstance(part, ThinkingPart):
+                        thinking_metadata: dict[str, Any] = {}
+                        if part.id is not None:
+                            thinking_metadata['id'] = part.id
+                        if part.signature is not None:
+                            thinking_metadata['signature'] = part.signature
+                        if part.provider_name is not None:
+                            thinking_metadata['provider_name'] = part.provider_name
+                        if part.provider_details is not None:
+                            thinking_metadata['provider_details'] = part.provider_details
+
+                        provider_metadata = {'pydantic_ai': thinking_metadata} if thinking_metadata else None
+                        ui_parts.append(
+                            ReasoningUIPart(text=part.content, state='done', provider_metadata=provider_metadata)
+                        )
+                    elif isinstance(part, FilePart):
+                        ui_parts.append(
+                            FileUIPart(
+                                url=part.content.data_uri,
+                                media_type=part.content.media_type,
+                            )
+                        )
+                    elif isinstance(part, BaseToolCallPart):
+                        if isinstance(part, BuiltinToolCallPart):
+                            prefixed_id = (
+                                f'{BUILTIN_TOOL_CALL_ID_PREFIX}|{part.provider_name or ""}|{part.tool_call_id}'
```

Comment:
We shouldn't need to make any special IDs, as we can store metadata on the part, right? And Vercel AI already has a native way of specifying `provider_executed` tool calls